GitCommands/Worktrees/WorktreeStatusProvider.cs

```diff
@@ -0,0 +1,134 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using GitUI;
+using GitUIPluginInterfaces;
+using Microsoft.VisualStudio.Threading;
+
+namespace GitCommands.Worktrees
+{
+    public interface IWorktreeStatusProvider : IDisposable
+    {
+        event EventHandler<WorktreeStatusEventArgs> StatusUpdated;
+
+        void Init();
+
+        Task UpdateWorktreesAsync(string workingDirectory);
+    }
+
+    public sealed class WorktreeStatusProvider : IWorktreeStatusProvider
+    {
+        private readonly CancellationTokenSequence _worktreesSequence = new CancellationTokenSequence();
+
+        private WorktreeInfoResult? _worktreeInfoResult;
+
+        // Singleton accessor
+        public static WorktreeStatusProvider Default { get; } = new WorktreeStatusProvider();
+
+        // Invoked when status update is requested (use to clear/lock UI)
+        public event EventHandler? StatusUpdating;
+
+        // Invoked when status update is complete
+        public event EventHandler<WorktreeStatusEventArgs>? StatusUpdated;
+
+        public void Dispose()
+        {
+            _worktreesSequence.Dispose();
+        }
+
+        public void Init()
+        {
+            // Cancel any previous async activities:
+            _worktreesSequence.Next();
+        }
+
+        public async Task UpdateWorktreesAsync(string workingDirectory)
+        {
+            _worktreeInfoResult = null;
+
+            // Cancel previous updates
+            var cancelToken = _worktreesSequence.Next();
+
+            OnStatusUpdating();
+
+            await TaskScheduler.Default;
+
+            // Start gathering worktrees asynchronously.
+            var currentModule = new GitModule(workingDirectory);
```

Comment:
I think all `var` in this file stand for pretty simple types which should be used explicitely.